core/src/main/java/org/apache/iceberg/encryption/KeyMetadata.java

```diff
@@ -0,0 +1,136 @@
+  private static final String wrappingKeyIdField = "wrapping_key_id";
+  private static final String aadPrefixField = "aad_prefix";
+  private static final byte V1 = 1;
+  private static final Schema SCHEMA_V1 =
+      new Schema(
+          required(0, encryptionKeyField, Types.BinaryType.get()),
+          optional(1, wrappingKeyIdField, Types.StringType.get()),
+          optional(2, aadPrefixField, Types.BinaryType.get()));
+  private static final ThreadLocal<KeyMetadataEncoder<IndexedRecord>> ENCODER =
+      ThreadLocal.withInitial(() -> new KeyMetadataEncoder<>(V1));
+  private static final ThreadLocal<KeyMetadataDecoder<IndexedRecord>> DECODER =
+      ThreadLocal.withInitial(() -> new KeyMetadataDecoder<>(V1));
+
+  static final Map<Byte, Schema> supportedSchemaVersions = ImmutableMap.of(V1, SCHEMA_V1);
+
+  private String wrappingKeyId;
+  private ByteBuffer encryptionKey;
+  private ByteBuffer aadPrefix;
+
+  KeyMetadata() {}
+
+  KeyMetadata(ByteBuffer encryptionKey, String wrappingKeyId, ByteBuffer aadPrefix) {
+    this.wrappingKeyId = wrappingKeyId;
+    this.encryptionKey = encryptionKey;
+    this.aadPrefix = aadPrefix;
+  }
+
+  String wrappingKeyId() {
+    return wrappingKeyId;
+  }
+
+  ByteBuffer encryptionKey() {
+    return encryptionKey;
+  }
+
+  ByteBuffer aadPrefix() {
+    return aadPrefix;
+  }
+
+  static KeyMetadata parse(ByteBuffer buffer) {
+    KeyMetadataDecoder<IndexedRecord> decoder = DECODER.get();
+
+    try {
+      return (KeyMetadata) decoder.decode(buffer);
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to parse envelope encryption metadata", e);
+    }
+  }
+
+  @Override
+  public ByteBuffer buffer() {
+    KeyMetadataEncoder<IndexedRecord> encoder = ENCODER.get();
+
+    try {
+      return encoder.encode(this);
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to serialize envelope key metadata", e);
+    }
+  }
+
+  @Override
+  public EncryptionKeyMetadata copy() {
+    KeyMetadata metadata = new KeyMetadata(encryptionKey(), wrappingKeyId(), aadPrefix());
+    return metadata;
+  }
+
+  @Override
+  public void put(int i, Object v) {
+    switch (i) {
+      case 0:
+        this.encryptionKey = (ByteBuffer) v;
+        return;
+      case 1:
+        this.wrappingKeyId = (v == null) ? null : v.toString();
+        return;
+      case 2:
+        this.aadPrefix = (ByteBuffer) v;
+        return;
+      default:
+        // ignore the object, it must be from a newer version of the format
+    }
+  }
+
+  @Override
+  public Object get(int i) {
+    switch (i) {
+      case 0:
+        return encryptionKey;
+      case 1:
+        return wrappingKeyId;
+      case 2:
+        return aadPrefix;
+      default:
+        throw new UnsupportedOperationException("Unknown field ordinal: " + i);
+    }
+  }
+
+  @Override
+  public org.apache.avro.Schema getSchema() {
+    return AvroSchemaUtil.convert(SCHEMA_V1, this.getClass().getCanonicalName());
```

Comment:
Rather than doing this every time the schema is requested, can you make it a static final variable?